// @tbiegner99/ui-app-components/vuild/ui-app-components.es.js
import _classCallCheck from '@babel/runtime/helpers/classCallCheck';
import _createClass from '@babel/runtime/helpers/createClass';
import _inherits from '@babel/runtime/helpers/inherits';
import _possibleConstructorReturn from '@babel/runtime/helpers/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/getPrototypeOf';
import axios from 'axios';
import _regeneratorRuntime from '@babel/runtime/regenerator';
import _asyncToGenerator from '@babel/runtime/helpers/asyncToGenerator';
import _defineProperty from '@babel/runtime/helpers/defineProperty';

var DispatcherFactory = /*#__PURE__*/function () {
  function DispatcherFactory() {
    _classCallCheck(this, DispatcherFactory);
  }

  _createClass(DispatcherFactory, null, [{
    key: "setDispatchingStrategy",
    value: function setDispatchingStrategy(dispatcher) {
      DispatcherFactory._dispatcher = dispatcher;
    }
  }, {
    key: "dispatcher",
    get: function get() {
      return DispatcherFactory._dispatcher;
    }
  }, {
    key: "dispatch",
    value: function dispatch(action) {
      DispatcherFactory.dispatcher.dispatch(action);
    }
  }]);

  return DispatcherFactory;
}();

var UrlEvents = {
  CHANGE_URL: 'CHANGE_URL'
};

var BaseActionCreator = /*#__PURE__*/function () {
  function BaseActionCreator() {
    _classCallCheck(this, BaseActionCreator);
  }

  _createClass(BaseActionCreator, [{
    key: "dispatch",
    value: function dispatch(action
    /* , data */
    ) {
      DispatcherFactory.dispatch(action);
    }
  }, {
    key: "changeUrl",
    value: function changeUrl(redirectLocation) {
      this.dispatch({
        type: UrlEvents.CHANGE_URL,
        data: {
          url: redirectLocation
        }
      });
    }
  }]);

  return BaseActionCreator;
}();

function _createSuper$1(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct$1(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct$1() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var ApplicationActionCreator = /*#__PURE__*/function (_BaseActionCreator) {
  _inherits(ApplicationActionCreator, _BaseActionCreator);

  var _super = _createSuper$1(ApplicationActionCreator);

  function ApplicationActionCreator() {
    _classCallCheck(this, ApplicationActionCreator);

    return _super.apply(this, arguments);
  }

  return ApplicationActionCreator;
}(BaseActionCreator);

var ApplicationActionCreator$1 = new ApplicationActionCreator();

var BASE_URL = '/api/kareoke';

var BaseDatasource = /*#__PURE__*/function () {
  function BaseDatasource(config) {
    var baseUrl = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : BASE_URL;

    _classCallCheck(this, BaseDatasource);

    this.baseUrl = baseUrl;
    this._client = axios.create(config);
  }

  _createClass(BaseDatasource, [{
    key: "constructUrl",
    value: function constructUrl(url) {
      return "".concat(this.baseUrl).concat(url);
    }
  }, {
    key: "client",
    get: function get() {
      return this._client;
    }
  }]);

  return BaseDatasource;
}();

var LoadingStates = {
  NOT_LOADED: 'NOT_LOADED',
  LOADING: 'LOADING',
  DONE: 'DONE',
  ERROR: 'ERROR'
};

var NOOP = function NOOP() {};

var StoreField = /*#__PURE__*/function () {
  function StoreField(name, defaultValue, onLoad) {
    _classCallCheck(this, StoreField);

    this._name = name;
    this._loadAction = onLoad || NOOP;
    this._loadState = defaultValue ? LoadingStates.DONE : LoadingStates.NOT_LOADED;
    this._value = defaultValue;
    this._error = null;
  }

  _createClass(StoreField, [{
    key: "error",
    get: function get() {
      return this._error;
    },
    set: function set(err) {
      this._error = err;
      this._loadState = LoadingStates.ERROR;
    }
  }, {
    key: "name",
    get: function get() {
      return this._name;
    }
  }, {
    key: "loadState",
    get: function get() {
      return this._loadState;
    }
  }, {
    key: "onLoad",
    set: function set(action) {
      this._loadAction = action;
    }
  }, {
    key: "loading",
    value: function loading() {
      this._loadState = LoadingStates.LOADING;
    }
  }, {
    key: "loadValue",
    value: function () {
      var _loadValue = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!(this.loadState !== LoadingStates.LOADING)) {
                  _context.next = 9;
                  break;
                }

                this._loadState = LoadingStates.LOADING;
                _context.prev = 2;
                _context.next = 5;
                return this._loadAction();

              case 5:
                _context.next = 9;
                break;

              case 7:
                _context.prev = 7;
                _context.t0 = _context["catch"](2);

              case 9:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 7]]);
      }));

      function loadValue() {
        return _loadValue.apply(this, arguments);
      }

      return loadValue;
    }()
  }, {
    key: "value",
    get: function get() {
      if (this.loadState !== LoadingStates.DONE) {
        this.loadValue();
      }

      return this._value;
    },
    set: function set(value) {
      this._value = value;
      this._loadState = LoadingStates.DONE;
    }
  }, {
    key: "valueIfNotLoading",
    get: function get() {
      if (this.loadState !== LoadingStates.DONE) {
        this.loadValue();
        return null;
      }

      return this._value;
    }
  }]);

  return StoreField;
}();

_defineProperty(StoreField, "LoadingStates", LoadingStates);

var AbstractReducingStore = /*#__PURE__*/function () {
  function AbstractReducingStore() {
    _classCallCheck(this, AbstractReducingStore);

    this.reduce = this.reduce.bind(this);
    this.handleEvent = this.handleEvent.bind(this);
  }

  _createClass(AbstractReducingStore, [{
    key: "reduce",
    value: function reduce(state, action) {
      var eventHandled = this.handleEvent(action);

      if (eventHandled || !state) {
        return {
          store: this
        };
      }

      return state;
    }
  }, {
    key: "handleEvent",
    value: function
      /* action */
    handleEvent() {}
  }]);

  return AbstractReducingStore;
}();

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var ApplicationStore = /*#__PURE__*/function (_AbstractReducingStor) {
  _inherits(ApplicationStore, _AbstractReducingStor);

  var _super = _createSuper(ApplicationStore);

  function ApplicationStore() {
    var _this;

    _classCallCheck(this, ApplicationStore);

    _this = _super.call(this);
    _this.data = {
      currentUrl: new StoreField('currentUrl', window.location.pathname)
    };
    return _this;
  }

  _createClass(ApplicationStore, [{
    key: "currentUrl",
    get: function get() {
      return this.data.currentUrl;
    }
  }, {
    key: "handleEvent",
    value: function handleEvent(action) {
      switch (action.type) {
        case UrlEvents.CHANGE_URL:
          this.data.currentUrl.value = action.data.url;
          break;

        default:
          return false;
      }

      return true;
    }
  }]);

  return ApplicationStore;
}(AbstractReducingStore);

var ApplicationStore$1 = new ApplicationStore();

export { AbstractReducingStore, ApplicationActionCreator$1 as ApplicationActionCreator, ApplicationStore$1 as ApplicationStore, BaseActionCreator, BaseDatasource, DispatcherFactory, LoadingStates, StoreField, UrlEvents };
