// @tbiegner99/ui-app-components/build/ui-app-components.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _classCallCheck = require('@babel/runtime/helpers/classCallCheck');
var _createClass = require('@babel/runtime/helpers/createClass');
var _inherits = require('@babel/runtime/helpers/inherits');
var _possibleConstructorReturn = require('@babel/runtime/helpers/possibleConstructorReturn');
var _getPrototypeOf = require('@babel/runtime/helpers/getPrototypeOf');
var axios = require('axios');
var _regeneratorRuntime = require('@babel/runtime/regenerator');
var _asyncToGenerator = require('@babel/runtime/helpers/asyncToGenerator');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _classCallCheck__default = /*#__PURE__*/_interopDefaultLegacy(_classCallCheck);
var _createClass__default = /*#__PURE__*/_interopDefaultLegacy(_createClass);
var _inherits__default = /*#__PURE__*/_interopDefaultLegacy(_inherits);
var _possibleConstructorReturn__default = /*#__PURE__*/_interopDefaultLegacy(_possibleConstructorReturn);
var _getPrototypeOf__default = /*#__PURE__*/_interopDefaultLegacy(_getPrototypeOf);
var axios__default = /*#__PURE__*/_interopDefaultLegacy(axios);
var _regeneratorRuntime__default = /*#__PURE__*/_interopDefaultLegacy(_regeneratorRuntime);
var _asyncToGenerator__default = /*#__PURE__*/_interopDefaultLegacy(_asyncToGenerator);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var DispatcherFactory = /*#__PURE__*/function () {
  function DispatcherFactory() {
    _classCallCheck__default["default"](this, DispatcherFactory);
  }

  _createClass__default["default"](DispatcherFactory, null, [{
    key: "setDispatchingStrategy",
    value: function setDispatchingStrategy(dispatcher) {
      DispatcherFactory._dispatcher = dispatcher;
    }
  }, {
    key: "dispatcher",
    get: function get() {
      return DispatcherFactory._dispatcher;
    }
  }, {
    key: "dispatch",
    value: function dispatch(action) {
      DispatcherFactory.dispatcher.dispatch(action);
    }
  }]);

  return DispatcherFactory;
}();

var UrlEvents = {
  CHANGE_URL: 'CHANGE_URL'
};

var BaseActionCreator = /*#__PURE__*/function () {
  function BaseActionCreator() {
    _classCallCheck__default["default"](this, BaseActionCreator);
  }

  _createClass__default["default"](BaseActionCreator, [{
    key: "dispatch",
    value: function dispatch(action
    /* , data */
    ) {
      DispatcherFactory.dispatch(action);
    }
  }, {
    key: "changeUrl",
    value: function changeUrl(redirectLocation) {
      this.dispatch({
        type: UrlEvents.CHANGE_URL,
        data: {
          url: redirectLocation
        }
      });
    }
  }]);

  return BaseActionCreator;
}();

function _createSuper$1(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct$1(); return function _createSuperInternal() { var Super = _getPrototypeOf__default["default"](Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf__default["default"](this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn__default["default"](this, result); }; }

function _isNativeReflectConstruct$1() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var ApplicationActionCreator = /*#__PURE__*/function (_BaseActionCreator) {
  _inherits__default["default"](ApplicationActionCreator, _BaseActionCreator);

  var _super = _createSuper$1(ApplicationActionCreator);

  function ApplicationActionCreator() {
    _classCallCheck__default["default"](this, ApplicationActionCreator);

    return _super.apply(this, arguments);
  }

  return ApplicationActionCreator;
}(BaseActionCreator);

var ApplicationActionCreator$1 = new ApplicationActionCreator();

var BASE_URL = '/api/kareoke';

var BaseDatasource = /*#__PURE__*/function () {
  function BaseDatasource(config) {
    var baseUrl = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : BASE_URL;

    _classCallCheck__default["default"](this, BaseDatasource);

    this.baseUrl = baseUrl;
    this._client = axios__default["default"].create(config);
  }

  _createClass__default["default"](BaseDatasource, [{
    key: "constructUrl",
    value: function constructUrl(url) {
      return "".concat(this.baseUrl).concat(url);
    }
  }, {
    key: "client",
    get: function get() {
      return this._client;
    }
  }]);

  return BaseDatasource;
}();

var LoadingStates = {
  NOT_LOADED: 'NOT_LOADED',
  LOADING: 'LOADING',
  DONE: 'DONE',
  ERROR: 'ERROR'
};

var NOOP = function NOOP() {};

var StoreField = /*#__PURE__*/function () {
  function StoreField(name, defaultValue, onLoad) {
    _classCallCheck__default["default"](this, StoreField);

    this._name = name;
    this._loadAction = onLoad || NOOP;
    this._loadState = defaultValue ? LoadingStates.DONE : LoadingStates.NOT_LOADED;
    this._value = defaultValue;
    this._error = null;
  }

  _createClass__default["default"](StoreField, [{
    key: "error",
    get: function get() {
      return this._error;
    },
    set: function set(err) {
      this._error = err;
      this._loadState = LoadingStates.ERROR;
    }
  }, {
    key: "name",
    get: function get() {
      return this._name;
    }
  }, {
    key: "loadState",
    get: function get() {
      return this._loadState;
    }
  }, {
    key: "onLoad",
    set: function set(action) {
      this._loadAction = action;
    }
  }, {
    key: "loading",
    value: function loading() {
      this._loadState = LoadingStates.LOADING;
    }
  }, {
    key: "loadValue",
    value: function () {
      var _loadValue = _asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee() {
        return _regeneratorRuntime__default["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!(this.loadState !== LoadingStates.LOADING)) {
                  _context.next = 9;
                  break;
                }

                this._loadState = LoadingStates.LOADING;
                _context.prev = 2;
                _context.next = 5;
                return this._loadAction();

              case 5:
                _context.next = 9;
                break;

              case 7:
                _context.prev = 7;
                _context.t0 = _context["catch"](2);

              case 9:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 7]]);
      }));

      function loadValue() {
        return _loadValue.apply(this, arguments);
      }

      return loadValue;
    }()
  }, {
    key: "value",
    get: function get() {
      if (this.loadState !== LoadingStates.DONE) {
        this.loadValue();
      }

      return this._value;
    },
    set: function set(value) {
      this._value = value;
      this._loadState = LoadingStates.DONE;
    }
  }, {
    key: "valueIfNotLoading",
    get: function get() {
      if (this.loadState !== LoadingStates.DONE) {
        this.loadValue();
        return null;
      }

      return this._value;
    }
  }]);

  return StoreField;
}();

_defineProperty__default["default"](StoreField, "LoadingStates", LoadingStates);

var AbstractReducingStore = /*#__PURE__*/function () {
  function AbstractReducingStore() {
    _classCallCheck__default["default"](this, AbstractReducingStore);

    this.reduce = this.reduce.bind(this);
    this.handleEvent = this.handleEvent.bind(this);
  }

  _createClass__default["default"](AbstractReducingStore, [{
    key: "reduce",
    value: function reduce(state, action) {
      var eventHandled = this.handleEvent(action);

      if (eventHandled || !state) {
        return {
          store: this
        };
      }

      return state;
    }
  }, {
    key: "handleEvent",
    value: function
      /* action */
    handleEvent() {}
  }]);

  return AbstractReducingStore;
}();

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf__default["default"](Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf__default["default"](this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn__default["default"](this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var ApplicationStore = /*#__PURE__*/function (_AbstractReducingStor) {
  _inherits__default["default"](ApplicationStore, _AbstractReducingStor);

  var _super = _createSuper(ApplicationStore);

  function ApplicationStore() {
    var _this;

    _classCallCheck__default["default"](this, ApplicationStore);

    _this = _super.call(this);
    _this.data = {
      currentUrl: new StoreField('currentUrl', window.location.pathname)
    };
    return _this;
  }

  _createClass__default["default"](ApplicationStore, [{
    key: "currentUrl",
    get: function get() {
      return this.data.currentUrl;
    }
  }, {
    key: "handleEvent",
    value: function handleEvent(action) {
      switch (action.type) {
        case UrlEvents.CHANGE_URL:
          this.data.currentUrl.value = action.data.url;
          break;

        default:
          return false;
      }

      return true;
    }
  }]);

  return ApplicationStore;
}(AbstractReducingStore);

var ApplicationStore$1 = new ApplicationStore();

exports.AbstractReducingStore = AbstractReducingStore;
exports.ApplicationActionCreator = ApplicationActionCreator$1;
exports.ApplicationStore = ApplicationStore$1;
exports.BaseActionCreator = BaseActionCreator;
exports.BaseDatasource = BaseDatasource;
exports.DispatcherFactory = DispatcherFactory;
exports.LoadingStates = LoadingStates;
exports.StoreField = StoreField;
exports.UrlEvents = UrlEvents;
